File: scripts/git-ai-after.js
#!/usr/bin/env node
/**
 * Git AI after-edit hook for DeepSeek TUI.
 * Called after edit_file / write_file / apply_patch tool completes.
 * Marks the just-written code as AI-authored via git-ai checkpoint agent-v1.
 *
 * Expects environment variables set by DeepSeek TUI's hook system:
 *   DEEPSEEK_TOOL_ARGS       — JSON of the tool call input
 *   DEEPSEEK_WORKSPACE       — git workspace root
 *   DEEPSEEK_MODEL           — model used (e.g. deepseek-v4-pro)
 *   DEEPSEEK_SESSION_ID      — current conversation/session id
 *   DEEPSEEK_TRANSCRIPT_PATH — path to transcript.json (optional)
 */

const { spawnSync } = require('child_process');
const fs = require('fs');

function extractFilePaths(toolArgsStr) {
  if (!toolArgsStr) return [];
  try {
    const args = JSON.parse(toolArgsStr);
    const paths = new Set();

    for (const key of ['path', 'file_path']) {
      const val = args[key];
      if (typeof val === 'string' && val) paths.add(val);
    }

    // apply_patch may have a "changes" array
    if (Array.isArray(args.changes)) {
      for (const change of args.changes) {
        if (change && typeof change === 'object') {
          const p = change.path || change.file_path;
          if (typeof p === 'string' && p) paths.add(p);
        }
      }
    }

    return [...paths];
  } catch {
    return [];
  }
}

function loadTranscript(transcriptPath) {
  if (!transcriptPath) return { messages: [] };
  try {
    const data = fs.readFileSync(transcriptPath, 'utf-8');
    const messages = JSON.parse(data);
    return { messages };
  } catch {
    return { messages: [] };
  }
}

function main() {
  const workspace = process.env.DEEPSEEK_WORKSPACE || '';
  const model = process.env.DEEPSEEK_MODEL || 'unknown';
  const sessionId = process.env.DEEPSEEK_SESSION_ID || 'unknown';
  const toolArgs = process.env.DEEPSEEK_TOOL_ARGS || '';
  const transcriptPath = process.env.DEEPSEEK_TRANSCRIPT_PATH;

  const filePaths = extractFilePaths(toolArgs);
  const transcript = loadTranscript(transcriptPath);

  const payload = JSON.stringify({
    type: 'ai_agent',
    repo_working_dir: workspace,
    agent_name: 'deepseek-tui',
    model,
    conversation_id: sessionId,
    edited_filepaths: filePaths,
    transcript,
  });

  const result = spawnSync('git-ai', ['checkpoint', 'agent-v1', '--hook-input', 'stdin'], {
    input: payload,
    encoding: 'utf-8',
    timeout: 30_000,
    stdio: ['pipe', 'inherit', 'inherit'],
  });

  if (result.error) {
    if (result.error.code === 'ENOENT') {
      process.stderr.write('git-ai: command not found\n');
    } else {
      process.stderr.write(`git-ai checkpoint failed: ${result.error.message}\n`);
    }
    process.exit(1);
  }
}

main();

File: scripts/git-ai-after.py
#!/usr/bin/env python3
"""
Git AI after-edit hook for DeepSeek TUI.
Called after edit_file / write_file / apply_patch tool completes.
Marks the just-written code as AI-authored via git-ai checkpoint agent-v1.

Expects environment variables set by DeepSeek TUI's hook system:
  DEEPSEEK_TOOL_ARGS       — JSON of the tool call input
  DEEPSEEK_WORKSPACE       — git workspace root
  DEEPSEEK_MODEL           — model used (e.g. deepseek-v4-pro)
  DEEPSEEK_SESSION_ID      — current conversation/session id
  DEEPSEEK_TRANSCRIPT_PATH — path to transcript.json (optional)
"""

import json
import os
import subprocess
import sys


def extract_file_paths(tool_args_str: str) -> list[str]:
    """Extract edited file paths from the tool call arguments."""
    if not tool_args_str:
        return []
    try:
        args = json.loads(tool_args_str)
    except json.JSONDecodeError:
        return []

    # Common field names for file paths across different tools
    paths = []
    for key in ("path", "file_path"):
        val = args.get(key)
        if val and isinstance(val, str):
            paths.append(val)

    # apply_patch may have a "changes" array or a "patch" string
    if "changes" in args and isinstance(args["changes"], list):
        for change in args["changes"]:
            if isinstance(change, dict):
                p = change.get("path") or change.get("file_path")
                if p and isinstance(p, str):
                    paths.append(p)

    return list(dict.fromkeys(paths))  # dedup preserving order


def load_transcript(transcript_path):
    """Load conversation transcript from the file path, if available."""
    if not transcript_path:
        return {"messages": []}
    if not os.path.isfile(transcript_path):
        return {"messages": []}
    try:
        with open(transcript_path, "r") as f:
            messages = json.load(f)
        # messages is already filtered (no tool results) by DeepSeek TUI
        return {"messages": messages}
    except (json.JSONDecodeError, OSError):
        return {"messages": []}


def main():
    workspace = os.environ.get("DEEPSEEK_WORKSPACE", "")
    model = os.environ.get("DEEPSEEK_MODEL", "unknown")
    session_id = os.environ.get("DEEPSEEK_SESSION_ID", "unknown")
    tool_args = os.environ.get("DEEPSEEK_TOOL_ARGS", "")
    transcript_path = os.environ.get("DEEPSEEK_TRANSCRIPT_PATH")

    file_paths = extract_file_paths(tool_args)
    transcript = load_transcript(transcript_path)

    payload = {
        "type": "ai_agent",
        "repo_working_dir": workspace,
        "agent_name": "deepseek-tui",
        "model": model,
        "conversation_id": session_id,
        "edited_filepaths": file_paths,
        "transcript": transcript,
    }

    try:
        proc = subprocess.run(
            ["git-ai", "checkpoint", "agent-v1", "--hook-input", "stdin"],
            input=json.dumps(payload),
            capture_output=False,
            text=True,
            timeout=30,
        )
    except FileNotFoundError:
        print("git-ai not found on PATH", file=sys.stderr, flush=True)
        sys.exit(1)
    except subprocess.TimeoutExpired:
        print("git-ai checkpoint timed out after 30s", file=sys.stderr, flush=True)
        sys.exit(1)


if __name__ == "__main__":
    main()
